fix(qbft): validate round change justification consistency with data_round

diegomrsantos opened the pull request:
## Issue Addressed

This PR improves QBFT round change message handling by implementing defensive validation that aligns with the EEA QBFT v1 specification's construction patterns.

## Proposed Changes

- Add defensive validation in `validate_proposal_justifications()` to ensure round change messages follow expected construction patterns
- Add corresponding validation in `received_round_change()` for consistency  
- Add comprehensive tests verifying the validation logic
- Complete documentation for the `validate_proposal_justifications` function

**Implementation Improvement**:
This validation implements defense-in-depth by ensuring:
1. Messages claiming no preparation (`data_round == 0`) have empty justifications (following spec construction intent)
2. Consistent message processing behavior across all validation paths
3. Prevention of unnecessary processing of irrelevant prepare messages

## Specification Alignment

### EEA QBFT v1 Construction Patterns

While the `validRoundChange` predicate only constrains payload fields (`preparedRound`/`preparedValue`), the specification's construction functions show the intended message shape:

- `createRoundChange` sets `roundChangeJustification` to `{}` when there's no prepared block
- `getRoundChangeJustification` returns `{}` for nodes with no preparation
- This establishes the expected construction pattern for honest implementations

### Proposal Validation Logic

The `isProposalJustification` predicate handles gating based on round change payload claims:
- If all RC payloads have no `preparedRound`/`preparedValue`, enters unprepared branch and ignores prepares
- Extra prepares in RCs with `preparedRound=None` are simply ignored during validation
- This defensive check prevents unnecessary processing without affecting correctness

### Implementation Rationale

This validation provides:
1. **Construction consistency**: Aligns implementation with specification's intended message construction
2. **Resource efficiency**: Prevents processing of irrelevant prepare messages
3. **Implementation robustness**: Removes potential confusion from inconsistent message patterns
4. **Defense-in-depth**: Additional validation layer without changing core protocol behavior

## Additional Info

This change represents a **defensive implementation improvement** that:
- Follows the specification's construction guidance more closely
- Prevents potential footguns in message processing
- Maintains full compatibility with the formal `validRoundChange` predicate requirements
- Adds no breaking changes to existing functionality

The validation mirrors the specification's construction patterns while providing additional implementation robustness.

diegomrsantos committed:
fix(qbft): validate round change justification consistency with data_round

Fix critical consensus vulnerability where malicious nodes could include unvalidated
prepare messages in round changes claiming no preparation (data_round=0).

Changes:
- Add validation in validate_proposal_justifications() to reject round changes
  with data_round=0 but non-empty prepare justifications
- Add validation in received_round_change() for the same vulnerability
- Add comprehensive test verifying both malicious and valid cases

According to EEA QBFT v1 specification:
- If data_round == 0: No prepare justifications should be present
- If data_round > 0: Prepare justifications MUST be validated

This prevents consensus safety violations where unvalidated prepare messages
could bypass validation and corrupt justification logic.

diegomrsantos committed:
docs(qbft): complete documentation comment for prepare justifications

- Clarify that prepare justifications are used to justify the highest prepared value from round changes
- Fix incomplete comment that was cut off mid-sentence

diegomrsantos committed:
fmt + lint

diegomrsantos committed:
Merge branch 'fix/qbft-round-change-justification-validation' of https://github.com/diegomrsantos/anchor into fix/qbft-round-change-justification-validation

Zacholme7 reviewed (dismissed):
Makes sense

Zacholme7 commented:
my bad, didnt see this was draft

diegomrsantos commented:
> my bad, didnt see this was draft

I changed to draft right now, I wanna give another review on this tomorrow with a fresh mind. But thanks anyway ;)

dknopik commented:
before merging this, let me test this against some fuzzer inputs I saved :pray:

diegomrsantos committed:
docs(qbft): fix test documentation placement

- Move doc comments above #[test] attributes for proper documentation
- Convert // comments to /// doc comments for test functions
- Ensure consistent documentation style across all tests

diegomrsantos committed:
refactor(qbft): eliminate code duplication in security test

- Extract helper functions to reduce test duplication:
  - create_signed_ssv_message(): Creates SSVMessage boilerplate
  - create_wrapped_round_change(): Creates round change messages
  - test_round_change_acceptance(): Tests acceptance/rejection logic
- Reduce test from ~200 lines to ~90 lines
- Improve maintainability and readability
- Same test coverage with cleaner code

diegomrsantos committed:
docs: add comprehensive code duplication prevention guidance

- Added DRY principle guidelines to Code Style section
- Added detailed Testing Guidelines with code duplication prevention
- Emphasized immediate refactoring when duplication is detected
- Added proper test documentation placement requirements
- Included helper function extraction patterns and examples

Prevents future code duplication issues by establishing clear
guidelines for when and how to extract helper functions.

diegomrsantos committed:
docs: add file size management for learning updates

- Added guidance to replace/consolidate rather than always add new content
- Emphasized concise, essential-only updates
- Added periodic review process for file maintenance
- Prioritized quality over quantity in guidelines

Prevents instruction files from becoming unwieldy over time.

diegomrsantos committed:
fmt

diegomrsantos committed:
docs: require format & lint before commits and remove auto-commit reference

- Added mandatory format & lint step before all commits
- Removed 'commit without permission' to respect user preferences
- Ensures code quality standards are maintained in learning updates

diegomrsantos committed:
refactor(docs): consolidate testing guidance and add proactive deduplication

- Streamline CLAUDE.md testing section to eliminate duplication with tester-subagent
- Add comprehensive proactive duplication prevention methodology to tester-subagent
- Enforce design-first approach: plan helper functions before writing tests
- Add enforcement rule: extract helpers immediately at 3+ line duplication
- Replace reactive cleanup approach with mandatory upfront pattern identification

diegomrsantos committed:
update wordlist.txt

diegomrsantos committed:
Merge branch 'release-v0.3.0' into fix/qbft-round-change-justification-validation

diegomrsantos commented on anchor/common/qbft/src/tests.rs:
```diff
@@ -356,3 +356,209 @@ fn test_round_change_validation_skips_round_one_prepared_values() {
+        root: Hash256::default(), // No preparation claimed
+        data_round: 0,            // Claims NO preparation
+        round_change_justification: vec![signed_malicious_prepare], // BUT includes justifications!
+        prepare_justification: vec![],
```
@dknopik how about this?

diegomrsantos committed:
update wordlist.txt

dknopik commented:
Related ssv spec issue: https://github.com/ssvlabs/ssv-spec/issues/582

It sounds like this is to be added as message validation rule (i.e. in `message_validator`). Wdyt about that?

diegomrsantos commented:
> Related ssv spec issue: [ssvlabs/ssv-spec#582](https://github.com/ssvlabs/ssv-spec/issues/582)
> 
> It sounds like this is to be added as message validation rule (i.e. in `message_validator`). Wdyt about that?

Isn't it different? `data_round == 0` and no justifications is consistent

diegomrsantos commented:
My initial idea was to implement this kind of validation in the QBFT crate, but I don't recall why I changed my mind. The advantage of having it in the validator is that we reject them faster and penalize the node, but not having this in the QBFT feels wrong.

dknopik commented:
>Isn't it different? data_round == 0 and no justifications is consistent

Oh, right, it is kinda different. 

I guess either all three or none of those should be true:

a) has full data
b) data_round != 0
c) has justifications

so it is kinda related?

dknopik commented:
> Without this validation, malicious nodes could:
(...)
3. Bypass consensus validation, potentially compromising safety

how?

diegomrsantos commented:
> > Isn't it different? data_round == 0 and no justifications is consistent
> 
> Oh, right, it is kinda different.
> 
> I guess either all three or none of those should be true:
> 
> a) has full data b) data_round != 0 c) has justifications
> 
> so it is kinda related?

Maybe. How does `full data` translate to the definition of `validRoundChange` on https://entethalliance.github.io/client-spec/qbft_spec.html?

diegomrsantos committed:
refactor(qbft): tone down test language

dknopik commented:
> Maybe. How does full data translate to the definition of validRoundChange on https://entethalliance.github.io/client-spec/qbft_spec.html?

I believe to `preparedValue: Optional<Hash>, ` within `UnsignedRoundChange`, but do not quote me on that.
